src/services/navigator/index.ts

```diff
@@ -0,0 +1,319 @@
+    let routeId = this.getRouteId(action, routeParams.url, hasPath);
+
+    let params:
+      | Types.NavigationNavigateParams
+      | TypesLegacy.NavigationRouteParams;
+    if (!path || !path.length) {
+      params = routeParams;
+    } else {
+      // The last path id is the screen id, remove from the path to avoid adding it in params
+      const lastPathId = path.pop();
+      params = this.buildParams(routeId, path, routeParams);
+      if (lastPathId) {
+        routeId = lastPathId;
+      }
+    }
+
+    return [navigation, routeId, params];
+  };
+
+  /**
+   * Process the request by changing params before going back
+   * Only the current navigator is targeted
+   * If the navigator is not type stack, the back request is bubbled
+   */
+  // eslint-disable-next-line class-methods-use-this
+  routeBackRequest(
+    navigation: HvRoute.RNTypedNavigationProps,
+    routeParams: TypesLegacy.NavigationRouteParams,
+  ) {
+    const state = navigation.getState();
+
+    if (state.type === Types.NAVIGATOR_TYPE.STACK && state.index > 0) {
+      const prev = state.routes[state.index - 1];
+
+      navigation.dispatch({
+        ...Imports.CommonActions.setParams({
+          ...routeParams,
+        }),
+        source: prev.key,
+        target: state.key,
+      });
+    }
+    navigation.goBack();
+  }
+
+  /**
+   * Prepare and send the request
+   */
+  sendRequest = (
+    action: TypesLegacy.NavAction,
+    routeParams: TypesLegacy.NavigationRouteParams,
+  ) => {
+    let { navigation } = this.props;
+    let routeId: string | undefined;
+    let params:
+      | Types.NavigationNavigateParams
+      | TypesLegacy.NavigationRouteParams
+      | undefined;
+    let navAction: TypesLegacy.NavAction = action;
+
+    if (routeParams) {
+      // The push action is used for urls which are not associated with a route
+      // See use of `this.getRouteKey(url);` in `hyperview/src/services/navigation`
+      if (routeParams.url) {
+        if (
+          navAction === TypesLegacy.NAV_ACTIONS.PUSH &&
+          Helpers.isUrlFragment(routeParams.url)
+        ) {
+          navAction = TypesLegacy.NAV_ACTIONS.NAVIGATE;
+        }
+      }
+
+      const [
+        requestNavigation,
+        requestRouteId,
+        requestParams,
+      ] = this.buildRequest(navAction, routeParams || {});
+
+      navigation = requestNavigation;
+      routeId = requestRouteId;
+      params = requestParams;
+    }
+
+    if (!navigation) {
+      if (routeParams.targetId) {
+        console.warn(
+          `No navigation found for target '${routeParams.targetId}'`,
+        );
+      }
+      return;
+    }
+
+    switch (navAction) {
+      case TypesLegacy.NAV_ACTIONS.BACK:
+        if (routeParams) {
+          this.routeBackRequest(navigation, routeParams);
+        } else {
+          navigation.goBack();
+        }
+        break;
+      case TypesLegacy.NAV_ACTIONS.CLOSE:
+        navigation.goBack();
```

Comment:
we should implement the routeParams stuff for the close action too

Reply:
Done in the new branch